### PR TITLE
feat(core): more type parsing

### PR DESCRIPTION
**Pull request by DaniPopes**
## Motivation

Upstreaming from foundry-cli/cast helper parsers

## PR Checklist

-   [ ] Added Tests
-   [ ] Added Documentation
-   [ ] Updated the changelog
-   [ ] Breaking changes

### CODE REVIEW

File: ethers-core/src/types/ens.rs [hunk 1]
```diff
@@ -100,10 +100,14 @@ impl From<Address> for NameOrAddress {
...
+    type Err = <Address as FromStr>::Err;
 
     fn from_str(s: &str) -> Result<Self, Self::Err> {
-        Ok(Self::Name(s.to_string()))
+        if s.starts_with("0x") {
+            s.parse().map(Self::Address)
+        } else {
+            Ok(Self::Name(s.to_string()))
+        }
     }
```
Comment (mattsse) [inline ethers-core/src/types/ens.rs:111]: this is a great change

File: ethers-core/src/types/block.rs [hunk 2]
```diff
@@ -624,15 +639,22 @@ impl FromStr for BlockNumber {
...
+            "latest" => Ok(Self::Latest),
+            "finalized" => Ok(Self::Finalized),
+            "safe" => Ok(Self::Safe),
+            "earliest" => Ok(Self::Earliest),
+            "pending" => Ok(Self::Pending),
+            n => {
+                if let Ok(n) = n.parse::<U64>() {
+                    Ok(Self::Number(n))
+                } else if let Ok(n) = n.parse::<u64>() {
+                    Ok(Self::Number(n.into()))
```
Comment (mattsse) [inline ethers-core/src/types/block.rs:652]: This is no longer compliant with the rpc spec for block number, but I guess this is fine?

**Review by mattsse — APPROVED** [review]
lgtm

File: ethers-core/src/types/block.rs (hunk 2 above)
Comment (gakonst) [reply ethers-core/src/types/block.rs:652]: Fine w/ me. More flexibility, and people create `BlockNumber` themselves, vs always parsing it out from the hex returned from RPC

**Review by gakonst — APPROVED** [review]

**Pull request merged**